### PR TITLE
fix(FASTQ-upload) Prevent all FASTQ uploads from failing if one fails

### PR DESCRIPTION
**Pull request by Vince-janv**
## Description
If the upload of FASTQ files fails on one case it will raise an error and stop the loop.

This PR catches any exceptions raised and continues the loop. But raises a click.abort error in the end if any case failed.

### Fixed

- One FASTQ upload fail no longer blocks other from being uploaded

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b catch-fastq-upload-errors
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/cli/upload/clinical_delivery.py [hunk 1]
```diff
@@ -83,12 +84,12 @@ def upload_clinical_delivery(context: click.Context, case_id: str, dry_run: bool
...
-
+    exit_code: int = EXIT_SUCCESS
     status_db: Store = context.obj.status_db
     trailblazer_api: TrailblazerAPI = context.obj.trailblazer_api
     for analysis_obj in status_db.get_analyses_to_upload(pipeline=Pipeline.FASTQ):
         if analysis_obj.family.analyses[0].uploaded_at:
-            LOG.warning(
+            LOG.debug(
                 "Newer analysis already uploaded for %s, skipping",
                 analysis_obj.family.internal_id,
```
Comment (diitaz93) [inline cg/cli/upload/clinical_delivery.py:94]:
```suggestion
            LOG.debug(
                f"Newer analysis already uploaded for {analysis_obj.family.internal_id}, skipping"
```

File: cg/cli/upload/clinical_delivery.py [hunk 2]
```diff
@@ -104,14 +105,24 @@ def auto_fastq(context: click.Context, dry_run: bool):
                 )
                 analysis_obj.uploaded_at = dt.datetime.now()
             else:
-                LOG.warning(
+                LOG.debug(
                     "Upload to clinical-delivery for %s has already started, skipping",
```
Comment (diitaz93) [inline cg/cli/upload/clinical_delivery.py:109]: same here

File: cg/cli/upload/clinical_delivery.py [hunk 3]
```diff
@@ -104,14 +105,24 @@ def auto_fastq(context: click.Context, dry_run: bool):
...
                     analysis_obj.family.internal_id,
                 )
             continue
         case: Family = analysis_obj.family
         LOG.info("Uploading family: %s", case.internal_id)
```
Comment (diitaz93) [inline cg/cli/upload/clinical_delivery.py:114]: f string

File: cg/cli/upload/clinical_delivery.py [hunk 4]
```diff
@@ -104,14 +105,24 @@ def auto_fastq(context: click.Context, dry_run: bool):
...
+        except Exception as error:
+            LOG.error(f"Upload of case {case.internal_id} failed")
+            LOG.error(error)
+            exit_code: int = EXIT_FAIL
+            continue
+
         if not dry_run:
             status_db.session.commit()
+
+        if exit_code:
```
Comment (diitaz93) [inline cg/cli/upload/clinical_delivery.py:127]: Im wondering if it should check for fail as `exit_code` is defined above as  `EXIT_SUCCESS`
```suggestion
        if exit_code == EXIT_FAIL:
```

**Review by diitaz93 — APPROVED** [review]
Looks good 👍

File: cg/cli/upload/clinical_delivery.py [hunk 5]
```diff
@@ -104,14 +105,24 @@ def auto_fastq(context: click.Context, dry_run: bool):
...
+        try:
+            context.invoke(upload_clinical_delivery, case_id=case.internal_id, dry_run=dry_run)
+        except Exception as error:
+            LOG.error(f"Upload of case {case.internal_id} failed")
+            LOG.error(error)
+            exit_code: int = EXIT_FAIL
+            continue
+
         if not dry_run:
             status_db.session.commit()
```
Comment (seallard) [inline cg/cli/upload/clinical_delivery.py:125]: This is interesting, it assumes that no commits are made outside of the CLI command layer (not true in general, maybe in this case). Probably a pattern we would like to implement as to tie the session/transaction life cycle to the duration of the CLI command

File: cg/cli/upload/clinical_delivery.py (hunk 4 above)
Comment (seallard) [reply cg/cli/upload/clinical_delivery.py:127]: Any non zero exit code indicates an error occurred, but maybe it is better to be explicit

File: cg/cli/upload/clinical_delivery.py [hunk 6]
```diff
@@ -104,14 +105,24 @@ def auto_fastq(context: click.Context, dry_run: bool):
...
                     "Upload to clinical-delivery for %s has already started, skipping",
                     analysis_obj.family.internal_id,
                 )
             continue
         case: Family = analysis_obj.family
         LOG.info("Uploading family: %s", case.internal_id)
         analysis_obj.upload_started_at = dt.datetime.now()
-        context.invoke(upload_clinical_delivery, case_id=case.internal_id, dry_run=dry_run)
+        try:
+            context.invoke(upload_clinical_delivery, case_id=case.internal_id, dry_run=dry_run)
```
Comment (seallard) [inline cg/cli/upload/clinical_delivery.py:117]: Not a fan of this pattern where we call other CLI commands, very much out of scope though 😅

**Review by seallard — APPROVED** [review]
Should work as intended 👍

File: cg/cli/upload/clinical_delivery.py [hunk 7]
```diff
@@ -104,14 +105,24 @@ def auto_fastq(context: click.Context, dry_run: bool):
...
+        except Exception as error:
+            LOG.error(f"Upload of case {case.internal_id} failed")
+            LOG.error(error)
+            exit_code: int = EXIT_FAIL
+            continue
```
Comment (islean) [inline cg/cli/upload/clinical_delivery.py:122]: Would we want some more explicit errors to catch?

File: cg/cli/upload/clinical_delivery.py (hunk 5 above)
Comment (islean) [inline cg/cli/upload/clinical_delivery.py:125]: I would also like to add that with this general commit, we could introduce errors with this change, if we commit changes for a failed upload. I suggest we either do a more specific error handling and do bot continue regardless of the Exception raised, or that we create a transaction for each case and either do multiple commits, or rollback transactions that fail and keep the "large" commit.

**Review by islean — CHANGES_REQUESTED** [review]
Somewhat worried about problems with database changes committed for failed uploads.

File: cg/cli/upload/clinical_delivery.py (hunk 4 above)
Comment (Vince-janv) [reply cg/cli/upload/clinical_delivery.py:127]: Changed to `if exit_code != EXIT_SUCCESS:`

File: cg/cli/upload/clinical_delivery.py (hunk 6 above)
Comment (Vince-janv) [reply cg/cli/upload/clinical_delivery.py:117]: Yep, it's the basis upon which cg is built 😢 If we can remove it in every module we re-write I'll be happy.

File: cg/cli/upload/clinical_delivery.py (hunk 7 above)
Comment (Vince-janv) [reply cg/cli/upload/clinical_delivery.py:122]: I don't think we want to. I see no type of error where we would not want to try with the next case 🤔  Do you see any potential bugs here?

File: cg/cli/upload/clinical_delivery.py (hunk 5 above)
Comment (Vince-janv) [reply cg/cli/upload/clinical_delivery.py:125]: Looking at the command we are calling, there are no changes to any ORM objects. But it's true that it the pattern is very prone to bugs. SA suggested we review it when looking at the session management with the latest SQL-alchemy version

File: cg/cli/upload/clinical_delivery.py (hunk 7 above)
Comment (islean) [reply cg/cli/upload/clinical_delivery.py:118]: Not really, I was thinking that maybe it is unnecessary to continue (and that an error should be raised) if we get a connection dropped error or something. But feel free to keep it as is.

File: cg/cli/upload/clinical_delivery.py (hunk 5 above)
Comment (islean) [reply cg/cli/upload/clinical_delivery.py]: Okay! Then I consider this comment to be out of scope :)

**Review by islean — APPROVED** [review]
Nice work! I think the pattern we use is somewhat error prone, but that is out of scope for this PR.

**Comment by Vince-janv** [comment]
Tested in stage.
The command did not exit after the first fail

**Comment by Vince-janv** [comment]
Deployed in production

**Pull request merged**